Incorrect type annotation on Generator.add_provider

* Faker version: 18.3.2
* OS: MacOS

The type annotation on `Generator.add_provider` is needlessly restrictive - the implementation clearly supports passing either a type or an instance, but the type annotation only allows instances:

```python
    # faker/generator.py
    def add_provider(self, provider: "BaseProvider") -> None:
        if isinstance(provider, type):
            provider = provider(self)
```


### Steps to reproduce

1. Write code that adds a provider class to a `Generator`
2. Run mypy

```python
class MyCustomProvider(BaseProvider):
    pass

fake = Factory.create("en_US")
fake.add_provider(MyCustomProvider)
```

### Expected behavior

`mypy` check should pass

### Actual behavior

`mypy` produces an error:

```
error: Argument 1 to "add_provider" of "Generator" has incompatible type "Type[MyCustomProvider]"; expected "BaseProvider"  [arg-type]
```

--- a/faker/generator.py
+++ b/faker/generator.py
@@ -1,7 +1,7 @@
 import random as random_module
 import re
 
-from typing import TYPE_CHECKING, Any, Callable, Dict, Hashable, List, Optional
+from typing import TYPE_CHECKING, Any, Callable, Dict, Hashable, List, Optional, Type, Union
 
 from .typing import SeedType
 
@@ -29,7 +29,7 @@ class Generator:
         self.__config = dict(list(self.__config.items()) + list(config.items()))
         self.__random = random
 
-    def add_provider(self, provider: "BaseProvider") -> None:
+    def add_provider(self, provider: Union["BaseProvider", Type["BaseProvider"]]) -> None:
         if isinstance(provider, type):
             provider = provider(self)
 

